app/controllers/InvigilatedController.scala

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2016 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import config.CSRCache
+import connectors.ApplicationClient
+import forms.VerifyCodeForm
+import helpers.NotificationType._
+import models.CachedData
+import play.api.mvc.{ Request, Result }
+
+import scala.concurrent.Future
+
+object InvigilatedController extends InvigilatedController(ApplicationClient, CSRCache)
+
+class InvigilatedController(applicationClient: ApplicationClient, cacheClient: CSRCache)
+  extends BaseController(applicationClient, cacheClient) {
+
+  def present = CSRUserAwareAction { implicit request =>
+    implicit user =>
+      Future.successful(Ok(views.html.index.invigilatedEtraySignin(VerifyCodeForm.form)))
+  }
+
+  def verifyToken = CSRUserAwareAction { implicit request =>
+    implicit user =>
+      VerifyCodeForm.form.bindFromRequest.fold(
+        invalidForm =>
+          Future.successful(Ok(views.html.index.invigilatedEtraySignin(invalidForm))),
+        data =>
+          applicationClient.verifyInvigilatedToken(data.email, data.token).flatMap {
+          invigilatedTest => Future.successful(Redirect(invigilatedTest.url))
+        }.fallbackTo(Future.successful(showValidationError(data)))
```

Comment:
it fallbacks all errors to one: validation error. I would add support for only one validation, and rest leave for common handler:

```
.recover {
        case e: YourException => Ok(views.html.application.assistanceDetails(yourEmptyForm))
      }
```